### ISSUE TITLE
How can I use generic component types when using styled api?

### ISSUE THREAD

**Issue by SoYoung210** [modulz/stitches#824]
# Bug report

## Describe the bug

when i use `styled`, can not receive the generic type that the component can accept.

```tsx
interface Props<T> {
  data?: T;
}

export function GenericComp<T extends string>({ data }: Props<T>) {
  console.log("data", data);
  return <div>generic component</div>;
}
```

actually, `GenericComp` can receive generic type like `<GenericComp<string> />`.  so when i use ..

```tsx
const StyledComp3 = styled<
  React.ComponentType<Props<number>>,
  (string | React.ComponentType<any> | Function | { [name: string]: unknown })[]
>(GenericComp, {
  marginTop: "4px"
});
```

i got new type error

```
Argument of type '{ marginTop: "4px"; }' is not assignable to parameter of type 'RemoveIndex<CSS<{}, {}, {}, {}>> & { variants?: { [x: string]: { [x: string]: CSS<{}, {}, {}, {}>; [x: number]: CSS<{}, {}, {}, {}>; }; } | undefined; compoundVariants?: (WideObject & { ...; })[] | undefined; defaultVariants?: WideObject | undefined; } & ... 5 more ... & { ...; }'.
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Go to [reproduce sandbox](https://codesandbox.io/s/eloquent-bas-23pmg?file=/src/App.tsx)
2. Click on `App.tsx`
3. hover at `StyledComp`

## Expected behavior

the component can receive generic type.

```tsx
// no type error!
<StyledComp<string> />
```

or just pass 1 generic prop at styled api

```tsx
// just one generic
const StyledComp3 = styled<
  React.ComponentType<Props<number>>,
>(GenericComp, {
  marginTop: "4px"
});

```

## Screenshots

![image](https://user-images.githubusercontent.com/18658235/135740813-2288cd10-4d9a-40e4-8a32-1b057b1ef574.png)

![image](https://user-images.githubusercontent.com/18658235/135741350-39897db9-1a92-43c8-a96e-9c1b66ae352e.png)

## System information

- OS: macOS
- Browser (if applies) chrome
- Version of Stitches: 1.2.1
- Version of Node.js:14

**Comment by hg-pyun** [comment]
I'm also curious about how to use it.

**Comment by peduarte** [comment]
Hmm, interesting. I dont fully understand the use case. I've never seen TS being used like this in JSX.

**Comment by always-maap** [comment]
> Hmm, interesting. I dont fully understand the use case. I've never seen TS being used like this in JSX.

Yeah, it's rare but there are examples on the internet. Passing select option type for example.
<Select<{value: number, label: string}> />

**Comment by SoYoung210** [comment]
> Hmm, interesting. I dont fully understand the use case. I've never seen TS being used like this in JSX.

I agree that it's not a common usage, but there are examples where this pattern is used in complex components (eg: Table). 😅

**Comment by hadihallak** [comment]
Will be closing this issue as this is not currently supported and we're not planing on supporting anytime soon as it's a bit challenging with how complex stitches types and the amount of edge cases we have to accommodate when it comes to typeings

**Comment by csantos1113** [comment]
> Will be closing this issue as this is not currently supported and we're not planing on supporting anytime soon as it's a bit challenging with how complex stitches types and the amount of edge cases we have to accommodate when it comes to typings

yeah, we also wanted to extend `styled` function to add some additional logic, but O.M.G.... impossible (same with `globalCss` and `css`)
